collector/impl/prometheus_metrics_collector.go

```diff
@@ -162,40 +165,96 @@ func (pc *PrometheusMetricsCollector) CollectMetrics() (metrics []hmetrics.Metri
+				Timestamp:      now,
+				Value:          g.GetValue(),
+				Tags:           labelPairMap,
+			}
+
+			metric := hmetrics.MetricHeader{
+				Type:   hmetrics.Gauge,
+				ID:     os.Expand(id, idMappingFunc),
+				Tenant: pc.Endpoint.Tenant,
+				Data:   data,
+			}
+
+			metricsCreated = append(metricsCreated, metric)
+		}
 
-	for i, m := range promMetricFamily.GetMetric() {
-		g := m.GetGauge()
-		metric.Data[i] = hmetrics.Datapoint{
-			Timestamp: now,
-			Value:     g.GetValue(),
-			Tags:      pc.prepareTagsMap(m.GetLabel()),
+	} else {
+		metric := hmetrics.MetricHeader{
+			Type:   hmetrics.Gauge,
+			ID:     id,
+			Tenant: pc.Endpoint.Tenant,
+			Data:   make([]hmetrics.Datapoint, len(promMetricFamily.GetMetric())),
 		}
+
+		for i, m := range promMetricFamily.GetMetric() {
+			g := m.GetGauge()
+			metric.Data[i] = hmetrics.Datapoint{
+				Timestamp: now,
+				Value:     g.GetValue(),
+				Tags:      pc.prepareTagsMap(m.GetLabel()),
+			}
+		}
+
+		metricsCreated = append(metricsCreated, metric)
 	}
 
 	return
 }
 
-func (pc *PrometheusMetricsCollector) convertCounter(promMetricFamily *prom.MetricFamily, id string, now time.Time) (metric hmetrics.MetricHeader) {
-	metric = hmetrics.MetricHeader{
-		Type:   hmetrics.Counter,
-		ID:     id,
-		Tenant: pc.Endpoint.Tenant,
-		Data:   make([]hmetrics.Datapoint, len(promMetricFamily.GetMetric())),
-	}
+func (pc *PrometheusMetricsCollector) convertCounter(promMetricFamily *prom.MetricFamily, id string, now time.Time) (metricsCreated []hmetrics.MetricHeader) {
+	metricsCreated = make([]hmetrics.MetricHeader, 0)
+	if strings.Contains(id, "$") {
+		for _, m := range promMetricFamily.GetMetric() {
+			labelPairMap := pc.prepareTagsMap(m.GetLabel())
+			idMappingFunc := expand.MappingFunc(false, labelPairMap)
+
+			// We still tag the data-points as not all labels have to be used on the metric name replacement
+			g := m.GetCounter()
+			data := make([]hmetrics.Datapoint, 1)
+			data[0] = hmetrics.Datapoint{
+				Timestamp:      now,
+				Value:          g.GetValue(),
+				Tags:           labelPairMap,
+			}
 
-	for i, m := range promMetricFamily.GetMetric() {
-		g := m.GetCounter()
-		metric.Data[i] = hmetrics.Datapoint{
-			Timestamp: now,
-			Value:     g.GetValue(),
-			Tags:      pc.prepareTagsMap(m.GetLabel()),
+			metric := hmetrics.MetricHeader{
+				Type:   hmetrics.Counter,
+				ID:     os.Expand(id, idMappingFunc),
+				Tenant: pc.Endpoint.Tenant,
+				Data:   data,
+			}
+
+			metricsCreated = append(metricsCreated, metric)
+		}
+
+	} else {
+		metric := hmetrics.MetricHeader{
+			Type:   hmetrics.Counter,
+			ID:     id,
+			Tenant: pc.Endpoint.Tenant,
+			Data:   make([]hmetrics.Datapoint, len(promMetricFamily.GetMetric())),
+		}
+
+		for i, m := range promMetricFamily.GetMetric() {
+			g := m.GetCounter()
+			metric.Data[i] = hmetrics.Datapoint{
+				Timestamp: now,
+				Value:     g.GetValue(),
+				Tags:      pc.prepareTagsMap(m.GetLabel()),
+			}
 		}
```

Comment:
Run with this PR and counters do not get stored. It is because here is missing the call to:

```
metricsCreated = append(metricsCreated, metric)
```